driver/src/orderbook/streamed/orderbook.rs

```diff
@@ -209,4 +209,55 @@ mod tests {
         // Cleanup the file created here.
         assert!(fs::remove_file(test_path).is_ok());
     }
+
+    #[test]
+    fn delete_events_starting_at_block() {
+        let mut orderbook = Orderbook::default();
+        for block_number in 0..3 {
+            orderbook.handle_event_data(
+                EventData::Added(Event::Deposit(Deposit::default())),
+                block_number,
+                0,
+                H256::zero(),
+                0,
+            );
+        }
+        assert_eq!(orderbook.events.len(), 3);
+        orderbook.delete_events_starting_at_block(1);
+        assert_eq!(orderbook.events.len(), 1);
+        assert_eq!(orderbook.events.iter().next().unwrap().0.block_number, 0);
+    }
+
+    #[test]
+    fn events_get_sorted() {
+        let mut orderbook = Orderbook::default();
+        // We are going to keep track of the correct order using the deposit's batch id.
+        let mut add_event = |deposit_id, block_number, log_index| {
+            orderbook.handle_event_data(
+                EventData::Added(Event::Deposit(Deposit {
+                    batch_id: deposit_id,
+                    ..Default::default()
+                })),
+                block_number,
+                log_index,
+                H256::zero(),
+                0,
+            );
+        };
+        // Add a couple of events in random order.
+        add_event(2, 0, 2);
+        add_event(4, 1, 1);
+        add_event(0, 0, 0);
+        add_event(5, 1, 2);
+        add_event(3, 1, 0);
+        add_event(1, 0, 1);
+        assert_eq!(orderbook.events.len(), 6);
+        // Check that the order is correct.
+        for (i, (_key, value)) in orderbook.events.iter().enumerate() {
```

Comment:
I find this test quite hard to understand. I have to wrap my had around this loop and dig into the implementation of the add_event closure to see deposit_id is used as batch_id. It's not very explicit that this is actually testing the sorting.

Can we maybe just list out the 6 value explicitly by calling .next() ?

Reply:
>  I'd prefer if the test cases could be more linear and easily to understand (even if it involves not writing the most elegant code, I think that is fine in tests).

I disagree. To me inlining and repeating everything makes the test harder to understand because there is more noise and I cannot be sure that all cases have been inlined the same way.

Reply:
Maybe a nice macro with a DSL might be a good compromise? While I also find noise in tests extremely frustrating (since they make the tests so long and hard to navigate because they are so many lines of similar code), maybe a macro with a dash of sugar can make things a bit easier to read?
```
add_deposit_events! {
        deposit # 2 block 0 log 2,
        deposit # 4 block 1 log 1,
        // ...
}
```

I find macros that define compact DSLs are a good way to make tests more readable and expressive without having too much additional noise.

Reply:
From a future maintainers perspective, I still don't fully understand what this is testing. 

We are checking that the batch_id is equal to the index of our loop. The connection between that and the thing that we are actually sorting by (block_number and event index) is super hidden in six randomly ordered  `add_event` with three times the same argument type combined with the detail that the deposit ID is reused as the batch_id (I like @nlordell's idea of using a macro that can name the types).

Moreover, isn't the fact that events are stored sorted in some way an implementation detail (it's not a promise we make via the interface). The class might as well store it as a Vec or HashMap. Therefore I'm not sure this should be tested. Maybe we want to test sorting `EventSortKey` directly but that doesn't require setting up an entire orderbook and events.

Reply:
> To me inlining and repeating everything makes the test harder to understand because there is more noise and I cannot be sure that all cases have been inlined the same way.

I do understand that long test cases are hard to read. However, unit tests should be testing logic and avoid adding more logic (which itself should be tested if it's non trivial). I agree it's a balance.

In this concrete example, writing 6 lines of the concrete expectation is in my mind much simpler to understand (worth 3 extra lines of code) than writing this enumerate loop.

Reply:
At this point the discussion about whether we should use loops and functions in tests isn't fruitful. We disagree on something subjective and the arguments have been said.

> Moreover, isn't the fact that events are stored sorted in some way an implementation detail (it's not a promise we make via the interface). The class might as well store it as a Vec or HashMap. Therefore I'm not sure this should be tested. Maybe we want to test sorting EventSortKey directly but that doesn't require setting up an entire orderbook and events.

It is an implementation detail. There is a philosophy of testing only public things and never implementation details. I disagree with taking this to the extreme but in general it makes sense.
In this specific case the alternative is to instead pick events that result in a valid `State` and then get the orderbook for that state to check that the events have been applied in the correct order. I considered this but didn't like that it would make the test more tedious and verbose.

I am going to change the test to work that way.

Reply:
What is DSL? Digital Second Line?

Reply:
domain specific language